File: java/src/main/java/com/tomoarrow/idv/bff/java/service/TokenService.java
package com.tomoarrow.idv.bff.java.service;

import com.tomoarrow.idv.bff.java.config.AppProperties;
import com.tomoarrow.idv.client.BodyOptions;
import com.tomoarrow.idv.client.ClientAssertionHelper;
import com.tomoarrow.idv.client.ClientAssertionOptions;
import com.tomoarrow.idv.client.generated.ApiException;
import com.tomoarrow.idv.client.generated.api.DefaultApi;
import com.tomoarrow.idv.client.generated.model.TokenRes;
import java.time.Instant;
import java.util.LinkedHashMap;
import java.util.Map;
import org.springframework.stereotype.Service;

@Service
public class TokenService {
    private final DefaultApi api;
    private final StateService stateService;
    private final AppProperties properties;

    public TokenService(DefaultApi api, StateService stateService, AppProperties properties) {
        this.api = api;
        this.stateService = stateService;
        this.properties = properties;
    }

    public TokenRes issueClientCredentialsToken() throws ApiException {
        if (isBlank(properties.getClientId()) || isBlank(properties.getSecret())) {
            throw new IllegalStateException("TOMO_IDV_CLIENT_ID and TOMO_IDV_SECRET must be set.");
        }

        String clientAssertion = ClientAssertionHelper.createClientAssertion(
                new ClientAssertionOptions(
                        properties.getClientId(),
                        properties.getSecret(),
                        properties.resolvedBaseUrl()
                )
        );

        TokenRes tokenResponse = api.v1Oauth2TokenPost(
                clientAssertion,
                BodyOptions.DEFAULT_CLIENT_ASSERTION_TYPE,
                BodyOptions.DEFAULT_GRANT_TYPE,
                BodyOptions.DEFAULT_RESOURCE,
                BodyOptions.DEFAULT_SCOPE
        );

        stateService.set("access_token", tokenResponse.getAccessToken());
        Map<String, Object> tokenInfo = new LinkedHashMap<>();
        tokenInfo.put("clientId", properties.getClientId());
        tokenInfo.put("tokenType", tokenResponse.getTokenType());
        tokenInfo.put("expiresIn", tokenResponse.getExpiresIn());
        tokenInfo.put("scope", tokenResponse.getScope());
        tokenInfo.put("issuedAt", Instant.now().toString());
        stateService.set("token_info", tokenInfo);
        return tokenResponse;
    }

    String bearerToken() {
        Object accessToken = stateService.get("access_token");
        if (!(accessToken instanceof String token) || token.isBlank()) {
            throw new IllegalStateException("No access token found. Please call /v1/oauth2/token first.");
        }
        return "Bearer " + token;
    }

    private boolean isBlank(String value) {
        return value == null || value.isBlank();
    }
}

File: kotlin/src/main/kotlin/com/tomoarrow/idv/bff/service/TokenService.kt
package com.tomoarrow.idv.bff.service

import com.tomoarrow.idv.bff.auth.JwtAssertionBuilder
import com.tomoarrow.idv.bff.config.AppProperties
import com.tomoarrow.idv.client.apis.DefaultApi
import org.springframework.stereotype.Service

/**
 * Handles OAuth2 client_credentials token issuance via JWT assertion.
 * Calls DefaultApi.v1Oauth2TokenPost() with the JWT assertion,
 * then stores the resulting access token in StateService.
 */
@Service
class TokenService(
    private val appProperties: AppProperties,
    private val stateService: StateService,
    private val api: DefaultApi
) {

    data class IssueAccessTokenResult(
        val clientId: String,
        val accessToken: String,
        val tokenType: String,
        val expiresIn: Long,
        val scope: String?
    )

    suspend fun issueClientCredentialsToken(): IssueAccessTokenResult {
        val baseUrl = appProperties.idvBaseUrl.trimEnd('/')
        val clientId = appProperties.tomoIdvClientId

        val clientAssertion = JwtAssertionBuilder.createClientAssertion(
            clientId = clientId,
            secretKey = appProperties.tomoIdvSecret,
            baseUrl = baseUrl
        )

        val tokenResponse = api.v1Oauth2TokenPost(
            clientAssertion = clientAssertion,
            clientAssertionType = "urn:ietf:params:oauth:client-assertion-type:jwt-bearer",
            grantType = "client_credentials",
            resource = "https://api.tomopayment.com/v1/idv",
            scope = "idv.read"
        )

        val accessToken = tokenResponse.accessToken ?: ""
        val tokenType = tokenResponse.tokenType ?: "Bearer"
        val expiresIn = tokenResponse.expiresIn ?: 0L

        stateService.set("access_token", accessToken)
        stateService.set("token_info", mapOf(
            "client_id" to clientId,
            "token_type" to tokenType,
            "expires_in" to expiresIn,
            "scope" to (tokenResponse.scope ?: ""),
            "issued_at" to java.time.Instant.now().toString()
        ))

        return IssueAccessTokenResult(
            clientId = clientId,
            accessToken = accessToken,
            tokenType = tokenType,
            expiresIn = expiresIn,
            scope = tokenResponse.scope
        )
    }

    fun requireAccessToken(): String {
        return stateService.get("access_token") as? String
            ?: throw IllegalStateException("No access token found. Please call /v1/oauth2/token first.")
    }
}
